src/webgpu/api/operation/command_buffer/programmable/programmable_state_test.ts

```diff
@@ -0,0 +1,223 @@
+    };
+
+    return this.device.createRenderPipeline({
+      layout: this.device.createPipelineLayout({
+        bindGroupLayouts: [this.bindGroupLayout, this.bindGroupLayout, this.bindGroupLayout],
+      }),
+      vertex: {
+        module: this.device.createShaderModule({
+          code: wgslShaders.vertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragment: {
+        module: this.device.createShaderModule({
+          code: wgslShaders.fragment,
+        }),
+        entryPoint: 'frag_main',
+        targets: [{ format: 'rgba8unorm' }],
+      },
+      primitive: { topology: 'point-list' },
+    });
+  }
+
+  beginSimplePass(passType: PassType): GPUProgrammablePassEncoder {
+    assert(this.encoder === null);
+    this.encoder = this.device.createCommandEncoder();
+    switch (passType) {
+      case PassType.Compute:
+        return this.encoder.beginComputePass();
+      case PassType.Render:
+        return this.beginSimpleRenderPass(this.encoder);
+      case PassType.RenderBundle:
+        return this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] });
+    }
+  }
+
+  beginSimpleRenderPass(encoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const view = this.device
+      .createTexture({
+        size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.RENDER_ATTACHMENT,
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+  }
+
+  setPipeline(pass: GPUProgrammablePassEncoder, pipeline: GPUComputePipeline | GPURenderPipeline) {
+    if (pass instanceof GPUComputePassEncoder) {
+      pass.setPipeline(pipeline as GPUComputePipeline);
+    } else if (pass instanceof GPURenderPassEncoder || pass instanceof GPURenderBundleEncoder) {
+      pass.setPipeline(pipeline as GPURenderPipeline);
+    }
+  }
+
+  dispatchOrDraw(pass: GPUProgrammablePassEncoder) {
+    if (pass instanceof GPUComputePassEncoder) {
+      pass.dispatch(1);
+    } else if (pass instanceof GPURenderPassEncoder) {
+      pass.draw(1);
+    } else if (pass instanceof GPURenderBundleEncoder) {
+      pass.draw(1);
+    }
+  }
+
+  endAndSubmitSimplePass(pass: GPUProgrammablePassEncoder) {
+    assert(this.encoder !== null);
+
+    if (pass instanceof GPUComputePassEncoder) {
+      pass.endPass();
+    } else if (pass instanceof GPURenderPassEncoder) {
+      pass.endPass();
+    } else if (pass instanceof GPURenderBundleEncoder) {
+      const renderBundle = pass.finish();
+      const renderPass = this.beginSimpleRenderPass(this.encoder);
+      renderPass.executeBundles([renderBundle]);
+      renderPass.endPass();
+    }
+
+    this.device.queue.submit([this.encoder.finish()]);
+    this.encoder = null;
+  }
+
+  verifyData(buffer: GPUBuffer, expectedValue: number) {
```

Comment:
nit: I think it would be more readable to simply do this inline instead of adding this method:
`this.expectGPUBufferValuesEqual(buffer, new Uint32Array([expectedValue]))`